### PR TITLE
[REG2.067a] Partial fix for issue 14133 - generate precise fillHoles code for struct literal expression

### PR DESCRIPTION
**Pull request by 9rnsr**
https://issues.dlang.org/show_bug.cgi?id=14133

The excessive fillHole code was introduced in #4387 to fix 13952, but originally I had to do this on that PR.

This PR will fix the regression case posted by Martin, but the general "excessive codegen" issue still exists around construction. Therefore after the merge, issue 14133 should be changed back to a normal issue.

### CODE REVIEW

File: src/e2ir.c [hunk 1]
```diff
@@ -5174,27 +5175,78 @@ elem *toElem(Expression *e, IRState *irs)
...
+                //printf("-- %s - fillHoles, structsize = %d\n", sle->toChars(), structsize);
+                for (size_t i = 0; i < sle->sd->fields.dim && offset < structsize; )
                 {
                     VarDeclaration *v = sle->sd->fields[i];
-                    size_t vend = v->offset + v->type->size();
-                    e = el_combine(e, fillHole(stmp, &offset, vend, sle->sd->structsize));
+                    if (i < dim && (*sle->elements)[i] && v->type->size())
+                    {
+                        //if (offset != v->offset) printf("  1 fillHole, %d .. %d\n", offset, v->offset);
+                        e = el_combine(e, fillHole(stmp, &offset, v->offset, structsize));
```
Comment (MartinNowak) [inline src/e2ir.c]: Looks like `StructDeclaration::fields` are ordered by offset, not sure how that works with overlapped fields.

**Comment by MartinNowak** [comment]
Well, it seems to works. It's pretty complex and I can't really review it.
It's remarkably similar to `StructDeclaration::fill`, why can't we use that one instead?

File: src/e2ir.c (hunk 1 above)
Comment (9rnsr) [inline src/e2ir.c]: `AggregateDeclaration::fields` is *not* ordered by offset, the fields are ordered by the lexical position in the code.

`VarDeclaration::overlapped == true` means that the variable is overlapped with other variables. For example:

union U { int a; int b; }
Both `a` and `b` will be set `overlapped = true`.

Comment (MartinNowak) [inline src/e2ir.c]: Ah, they are added in lexical order.
https://github.com/D-Programming-Language/dmd/blob/0701de6dd84ca7c9998ed41f50bf33d626082493/src/struct.c#L960

**Comment by 9rnsr** [comment]
> It's remarkably similar to StructDeclaration::fill, why can't we use that one instead?

`StructDeclaration::fill` works for the unspecified struct fields. This PR works for the holes between specified fields. Yes' they're similar because both should handle overlapped fields, but the holes are not handled in semantic analysis stage. So the former cannot handle the latter.

File: src/e2ir.c (hunk 1 above)
Comment (9rnsr) [inline src/e2ir.c]: Yes, and built-in `tupleof` and others depend on the order of `fields`. Changing it will be problematic.

File: src/e2ir.c [hunk 2]
```diff
@@ -5174,27 +5175,78 @@ elem *toElem(Expression *e, IRState *irs)
...
+                const size_t structsize = sle->sd->structsize;
                 size_t offset = 0;
-                for (size_t i = 0; i < sle->sd->fields.dim; i++)
+                //printf("-- %s - fillHoles, structsize = %d\n", sle->toChars(), structsize);
+                for (size_t i = 0; i < sle->sd->fields.dim && offset < structsize; )
                 {
                     VarDeclaration *v = sle->sd->fields[i];
-                    size_t vend = v->offset + v->type->size();
-                    e = el_combine(e, fillHole(stmp, &offset, vend, sle->sd->structsize));
+                    if (i < dim && (*sle->elements)[i] && v->type->size())
```
Comment (MartinNowak) [inline src/e2ir.c]: I don't understand what the initializer elements have to do with filling alignment holes.
Does this try to use the fact that you can't initialize multiple overlapping fields?
Some hint if form of an assertion or comment would be appropriate.

**Comment by MartinNowak** [comment]
Auto-merge toggled on

**Comment by MartinNowak** [comment]
Let me pull this, but we should find out how to simplify this. How about using a bit vector with one bit for each struct bytes. Then run over all (initialized?) fields marking the bits and afterwards generate code to fill the holes?
Also why do we need to fill alignment holes at all? Just so that we can use memcmp for struct comparison or compute a hash over the raw data?

**Comment by 9rnsr** [comment]
Thanks for the merge.

> How about using a bit vector with one bit for each struct bytes.

I thought it would need too much memory space if a struct is too big (e.g. `struct S { int[16 * 1024] buf; }`).

> Also why do we need to fill alignment holes at all? Just so that we can use memcmp for struct comparison

It's just for that.

File: src/e2ir.c (hunk 2 above)
Comment (9rnsr) [inline src/e2ir.c]: I opened PR #4509 to add comments.

**Comment by MartinNowak** [comment]
Well 2K isn't that much and one could reuse the buffers. But it would be a lot slower for big structs to set all the bits.

> It's just for that.

Might be interesting to run a few benchmarks, whether memcmp is worth it.

**Pull request merged**